_posts/2020-12-27-spring-boot-exception-hanlding.md

```diff
@@ -0,0 +1,596 @@
+        super(HttpStatus.NOT_FOUND, message);
+    }
+}
+```
+
+This approach comes in handy when we want to manipulate the response headers too.
+
+Now, as we can see `@ResponseStatus` along with `server.error` properties allows us to manipulate almost all the fields
+in our Spring defined error response payload but, what if want to manipulate response payload structure too? Let's see how
+we can achieve that in the next section. 
+
+## `@ExceptionHandler`
+`@ExceptionHandler` annotation gives us a lot of flexibility in terms of handling exceptions. For starters, to use it, we
+simply need to create a method either in the controller itself or in a `@ControllerAdvice` class and
+annotate it with `@ExceptionHandler`.
+
+```java
+@RestController
+@RequestMapping("/product")
+public class ProductController {
+    private final ProductService productService;
+    //constructor omitted for brevity...
+
+    @GetMapping("/{id}")
+    public Response getProduct(@PathVariable String id) {
+        return productService.getProduct(id);
+    }
+
+    @ExceptionHandler(NoSuchElementFoundException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ResponseEntity<String> handleNoSuchElementFoundException(
+            NoSuchElementFoundException exception
+    ) {
+        return ResponseEntity
+                .status(httpStatus)
+                .body(exception.getMessage());
+    }
+
+}
+```
+
+`@ExceptionHandler` takes in an exception, or a list of exceptions as an argument that we want to handle in the defined
+method. If we don't wish to do that then simply defining exception as a parameter of the method will also do:
+
+```java
+@ExceptionHandler
+public ResponseEntity<String> handleNoSuchElementFoundException(
+        NoSuchElementFoundException exception
+        )
+```
+
+Or we can keep both just as shown in the example controller above. It gives better readability. For the same reason, we have
+kept `@ResponseStatus(HttpStatus.NOT_FOUND)` on the handler method. Although it's not required as we have already provided HTTP status in `ResponseEntity`.
+
+Apart from exception parameter we can also have `HttpServletRequest`, `WebRequest`, `HttpSession` types as parameters. Similarly, the handler
+methods support a variety of return types such as `ResponseEntity`, `String`, or even `void`.
+Find more input and return types in [`@ExceptionHandler` java documentation](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/ExceptionHandler.html).
+
+With the plethora of options available to us in form of both input parameters and return types in our exception handling function,
+we are in complete control of the error response.
+
+Now, let's finalize an error response payload for our APIs. In case of any error clients usually expect two things:
+* First is error code which tells what kind of error it is. Also, error codes can be used by clients in their code to drive
```

Comment:
This list is introduced with "two things", but it has 4 bullet points. That's a bit too unexpected for my taste :)